### PR TITLE
[C/PyTorch] Simplify THD offset tensors

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR reduces the THD offset tensors from four (`seq_offsets_q`, `seq_offsets_k`, `seq_offsets_v`, `seq_offsets_o`) to two (`cu_seqlens_q_padded`, `cu_seqlens_kv_padded`).

Before this PR, for THD_THD_THD layout, users need to calculate these four tensors:
```
seq_offsets_q =           config.num_heads * config.head_dim * cu_seqlens_q_padded
seq_offsets_k = config.num_gqa_groups * config.head_dim * cu_seqlens_kv_padded
seq_offsets_v = config.num_gqa_groups * config.head_dim * cu_seqlens_kv_padded
seq_offsets_o =            config.num_heads * config.head_dim * cu_seqlens_q_padded
```
With this PR, users only need to provide two tensors, `cu_seqlens_q_padded` and `cu_seqlens_kv_padded`, which are easier to understand and utilize correctly.

An example of the difference between `cu_seqlens` and `cu_seqlens_padded` is, for a batch `[a, PAD, b, b, c, PAD, PAD, d, d]`, we have 4 sequences, `cu_seqlens = [0, 1, 3, 4, 6]`, and `cu_seqlens_padded= [0, 2, 4, 7, 9]`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Reduced the THD offset tensors from four to two
- This is breaking change compared to #832

# Checklist:

- [ x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ x] The functionality is complete
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] I have made corresponding changes to the documentation
- [ x] My changes generate no new warnings
- [ x] I have added tests that prove my fix is effective or that my feature works
- [ x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cyanguwa** [comment]
Hi @zlsh80826 ,

Yes, this PR is just focused on item 3. I wanted to get this done first so there is no API change between v1.8 and v1.9. I'm still evaluating the benefits/changes for items 1 and 2, but they will cause breaking API changes anyway. There is no urgency to it (well, not as much as item 3, given the code freeze coming up).

Thanks.

**Review by zlsh80826 — APPROVED** [review]

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 1]
```diff
@@ -196,21 +196,16 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
...
 void nvte_fused_attn_fwd_qkvpacked(const NVTETensor QKV, const NVTETensor Bias, NVTETensor S,
                                    NVTETensor O, NVTETensorPack *Aux_CTX_Tensors,
-                                   const NVTETensor cu_seqlens, const NVTETensor seq_offsets_q,
-                                   const NVTETensor seq_offsets_k, const NVTETensor seq_offsets_v,
-                                   const NVTETensor seq_offsets_o, const NVTETensor rng_state,
-                                   size_t max_seqlen, bool is_training, float attn_scale,
-                                   float dropout, NVTE_QKV_Layout qkv_layout,
+                                   const NVTETensor cu_seqlens, const NVTETensor cu_seqlens_padded,
+                                   const NVTETensor rng_state, size_t max_seqlen, bool is_training,
+                                   float attn_scale, float dropout, NVTE_QKV_Layout qkv_layout,
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:201]: I will address this in another PR as discussed.

File: transformer_engine/common/fused_attn/utils.cu [hunk 2]
```diff
@@ -360,6 +360,38 @@ __global__ void cu_seqlens_to_actual_seqlens(size_t b, int32_t const *const q_cu
     kv_seqlens[tid] = kv_cu_seqlens[tid + 1] - kv_cu_seqlens[tid];
   }
 }
+
+// convert cu_seqlens_padded to offsets
+__global__ void cu_seqlens_padded_to_offsets(NVTE_QKV_Layout_Group layout_group, size_t b, size_t h,
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/utils.cu:365]: I will address this in another PR.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h [hunk 3]
```diff
@@ -22,17 +22,15 @@ void fused_attn_arbitrary_seqlen_fwd_qkvpacked(
     size_t batch, size_t num_attn_heads, size_t max_seqlen, size_t head_dim, bool is_training,
     float attn_scale, float p_dropout, NVTE_QKV_Layout qkv_layout, NVTE_Bias_Type bias_type,
     NVTE_Mask_Type mask_type, const Tensor *input_QKV, const Tensor *input_Bias, Tensor *output_O,
-    NVTETensorPack *Aux_CTX_Tensors, const Tensor *cu_seqlens, const Tensor *seq_offsets_q,
-    const Tensor *seq_offsets_k, const Tensor *seq_offsets_v, const Tensor *seq_offsets_o,
+    NVTETensorPack *Aux_CTX_Tensors, const Tensor *cu_seqlens, const Tensor *cu_seqlens_padded,
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.h:25]: I tried a few different names as you can see from the commits, `cu_seqlens_with_offset`, `cu_seqlens_with_padding`, `cu_seqlens_padded`. They either became too long (with the `_q_/_kv_` added), or very confusing about what they mean. `seq_offsets_q/_kv` have the word `offset` in it but they loose the connection with `cu_seqlens`. Essentially, we are asking users to supply another kind of `cu_seqlens` that accounts for the padding between sequences.

**Pull request merged**